File: RegexNfa/Infrastructure/State.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace RegexNfa.Infrastructure
{
    public class State
    {
        public string Id { get; }
        public bool Accepting { get; set; }
        public Dictionary<string, Transition> FromTransitions;
        public Dictionary<string, Transition> ToTransitions;

        public State()
        {
            Id = Guid.NewGuid().ToString();
            Accepting = false;
            FromTransitions = new Dictionary<string, Transition>();
            ToTransitions = new Dictionary<string, Transition>();
        }

        public State(bool accepting) :
            this()
        {
            Accepting = accepting;
        }

        public void AddFromTransition(Transition transition)
        {
            try
            {
                FromTransitions.Add(transition.Id, transition);
            } catch (ArgumentException ex)
            {
                return;
            }

        }

        public void AddToTransition(Transition transition)
        {
            try
            {
                ToTransitions.Add(transition.Id, transition);
            } catch (ArgumentException ex)
            {
                return;
            }
        }

        public JObject ToJson()
        {
            JObject json = new JObject(
                new JProperty("Id", Id),
                new JProperty("Accepting", Accepting));
            return json;
        }

        public override string ToString()
        {
            return this.ToJson().ToString();
        }

    }
}

File: src/state.rs
use super::transition::Transition;

#[derive(Debug)]
pub struct State<'a> {
    accepting: bool,
    from_transitions: Vec<&'a Transition<'a>>,
    to_transitions: Vec<&'a Transition<'a>>,
}

impl<'a> State<'a> {
    pub fn new(accepting: bool) -> State<'a> {
        State {
            accepting,
            from_transitions: Vec::new(),
            to_transitions: Vec::new(),
        }
    }

    pub fn add_from_transition(&mut self, from_transition: &'a Transition) {
        self.from_transitions.push(from_transition);
    }
}
